### PR TITLE
Create Android Application plugin

### PR DESCRIPTION
**Pull request by jvandort**
Adds an Android Application plugin alongside the Android Library plugin. Much of the logic and data types have been copied, with slight modifications. All data types were simplified to not require any concrete implementations. The entire model can be expressed as interfaces that Gradle can generate implementations for.

Furthermore, we use build type terminology everywhere, as targets are generally not accessible in a static context. Targets, or variants in android naming, are calculated by AGP in an after evaluate and are therefore fundamentally inaccessible from the declarative schemas without some sort of dynamic validation.

### CODE REVIEW

File: unified-prototype/unified-plugin/plugin-common/src/main/java/org/gradle/api/experimental/common/ApplicationDependencies.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import org.gradle.api.plugins.jvm.PlatformDependencyModifiers;
+import org.gradle.api.plugins.jvm.TestFixturesDependencyModifiers;
+import org.gradle.declarative.dsl.model.annotations.Restricted;
+
+/**
+ * The declarative dependencies DSL block for an application.
+ */
+@SuppressWarnings("UnstableApiUsage")
+@Restricted
+public interface ApplicationDependencies extends PlatformDependencyModifiers, TestFixturesDependencyModifiers, GradleDependencies {
```
Comment (tresat) [inline unified-prototype/unified-plugin/plugin-common/src/main/java/org/gradle/api/experimental/common/ApplicationDependencies.java:14]: 🤔 I think we should just extend `Dependencies` directly until we get the modifiers from these other types working. 

Also on the `LibraryDependencies`.

File: unified-prototype/unified-plugin/plugin-common/src/main/java/org/gradle/api/experimental/common/ApplicationDependencies.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    DependencyCollector getImplementation();
```
Comment (tresat) [inline unified-prototype/unified-plugin/plugin-common/src/main/java/org/gradle/api/experimental/common/ApplicationDependencies.java:15]: 💭 Does it make sense to still call this "implementation" for an application?  It probably shouldn't revert to "compile" (or should it)?  Is this an opportunity to come up with a new application-only name here and leave impl/api for libraries only?

File: unified-prototype/README.md [hunk 3]
```diff
@@ -53,28 +53,28 @@ cd testbed-kmp
 
 ## Android
 
-The sample Android project lives in the `testbed-android` directory and has been updated to use the Declarative Gradle DSL. 
+The sample Android projects lives in the `testbed-android-library` and `testbed-android-application` directories. They have been updated to use the Declarative Gradle DSL.
```
Comment (tresat) [inline unified-prototype/README.md:56]:
```suggestion
Sample Android projects live in the `testbed-android-library` and `testbed-android-application` directories. They have been updated to use the Declarative Gradle DSL. 
```

File: unified-prototype/testbed-android-application/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (tresat) [inline unified-prototype/testbed-android-application/src/main/AndroidManifest.xml:1]: ❓ Shouldn't this live in a `main/resources` dir and not directly under `main`?

File: unified-prototype/unified-plugin/plugin-android/src/main/java/org/gradle/api/experimental/android/AndroidDSLSupport.java [hunk 5]
```diff
@@ -16,109 +16,15 @@
...
 import org.gradle.api.provider.Property;
-import org.jetbrains.kotlin.gradle.dsl.KotlinAndroidProjectExtension;
 
-import javax.annotation.Nullable;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Objects;
-import java.util.Set;
-
-public class AndroidDSLSupport {
```
Comment (tresat) [inline unified-prototype/unified-plugin/plugin-android/src/main/java/org/gradle/api/experimental/android/AndroidDSLSupport.java:33]: ❓ Why remove the common utilities into 2 separate classes?  This code is identical or very similar between libraries and applications.

File: unified-prototype/testbed-android-application/src/main/AndroidManifest.xml (hunk 4 above)
Comment (jvandort) [reply unified-prototype/testbed-android-application/src/main/AndroidManifest.xml:1]: Apparently not. When I move it I get:
```
Execution failed for task ':testbed-android-application:createDebugCompatibleScreenManifests'.
> Error while evaluating property 'variantOutputs.$0.versionCode' of task ':testbed-android-application:createDebugCompatibleScreenManifests'.
   > Manifest file does not exist: /Users/jvandort/work/declarative-prototypes/unified-prototype/testbed-android-application/src/main/AndroidManifest.xml
```

File: unified-prototype/unified-plugin/plugin-android/src/main/java/org/gradle/api/experimental/android/AndroidDSLSupport.java (hunk 5 above)
Comment (jvandort) [reply unified-prototype/unified-plugin/plugin-android/src/main/java/org/gradle/api/experimental/android/AndroidDSLSupport.java:33]: 1. They were not really common at all. The methods were only called in one place so I didn't see a reason to pull them out to another class. 
2. While both plugins use very similar logic, they cannot share logic due to the separate type hierarchies. If we end up merging the type hierarchies in the future I can see value in pulling out the logic to a common location but as it stands now each separate implementation of these linking methods are only called from their respective plugin.

**Pull request merged**